common/src/main/java/io/pravega/common/util/AbstractDrainingQueue.java

```diff
@@ -147,6 +152,41 @@ public void add(T item) {
         }
     }
 
+    /**
+     * Returns the next items from the queue. If the queue is empty, it blocks the call until at least one item is added,
+     * or until the given timeout expires.
+     *
+     * @param maxCount        The maximum number of items to return. This argument will be ignored if the queue is currently
+     *                        empty, but in that case the result will always be completed with exactly one element.
+     * @param timeout         Timeout.
+     * @param timeoutExecutor An Executor to use for timing out.
+     * @return A CompletableFuture that, when completed, will contain the requested result. If the queue is not currently
+     * empty, this Future will already be completed, otherwise it will be completed the next time the add() method is called.
+     * If the queue is closed and this Future is not yet completed, it will be cancelled. If the timeout expires prior
+     * to adding a new element, this future will be completed with a {@link TimeoutException}.
+     * @throws ObjectClosedException If the Queue is closed.
+     * @throws IllegalStateException If another call to take() is in progress.
+     */
+    public CompletableFuture<Queue<T>> take(int maxCount, Duration timeout, ScheduledExecutorService timeoutExecutor) {
+        CompletableFuture<Queue<T>> result = take(maxCount);
+        if (!result.isDone()) {
+            ScheduledFuture<?> sf = timeoutExecutor.schedule(() -> {
+                synchronized (this.lock) {
+                    if (this.pendingTake == result) {
+                        this.pendingTake = null;
+                    }
+                }
+
+                // Timeout the future after we have unregistered it. A zealous callback may invoke us immediately and be
+                // surprised that their call is rejected since we can only have one take() outstanding at any given time.
+                result.completeExceptionally(new TimeoutException());
+            }, timeout.toMillis(), TimeUnit.MILLISECONDS);
+            result.whenComplete((r, ex) -> sf.cancel(true));
```

Comment:
Yes, I am aware of that. All of our executors have that flag set (see `ExecutorServiceHelpers.java`) precisely for this reason. 

FYI, this code is borrowed from `Futures.futureWithTimeout`. I didn't want to reuse that method since I wanted the leanest possible implementation which was tied into the `take()` future above.